src/serac/physics/solid_mechanics_contact.hpp

```diff
@@ -0,0 +1,353 @@
+      const mfem::Vector res =
+          (*residual_)(u_blk, zero_, shape_displacement_, *parameters_[parameter_indices].state...);
+
+      // TODO this copy is required as the sundials solvers do not allow move assignments because of their memory
+      // tracking strategy
+      // See https://github.com/mfem/mfem/issues/3531
+      mfem::Vector r_blk(r, 0, displacement_.Size());
+      r_blk = res;
+      contact_.residualFunction(u, r);
+      // we need to do this again to clear out contact residual on essential dofs
+      r_blk.SetSubVector(bcs_.allEssentialTrueDofs(), 0.0);
+    };
+    if (contact_.haveLagrangeMultipliers()) {
+      J_offsets_ = mfem::Array<int>({0, displacement_.Size(), displacement_.Size() + contact_.numPressureDofs()});
+      return std::make_unique<mfem_ext::StdFunctionOperator>(
+          displacement_.space().TrueVSize() + contact_.numPressureDofs(), residual_fn,
+          // gradient of residual function
+          [this](const mfem::Vector& u) -> mfem::Operator& {
+            const mfem::Vector u_blk(const_cast<mfem::Vector&>(u), 0, displacement_.Size());
+            auto [r, drdu] = (*residual_)(differentiate_wrt(u_blk), zero_, shape_displacement_,
+                                          *parameters_[parameter_indices].state...);
+            J_             = assemble(drdu);
+
+            // create block operator holding jacobian contributions
+            J_constraint_ = contact_.jacobianFunction(u, J_.release());
+
+            // take ownership of blocks
+            J_constraint_->owns_blocks = false;
+            J_                         = std::unique_ptr<mfem::HypreParMatrix>(
+                static_cast<mfem::HypreParMatrix*>(&J_constraint_->GetBlock(0, 0)));
+            J_12_ = std::unique_ptr<mfem::HypreParMatrix>(
+                static_cast<mfem::HypreParMatrix*>(&J_constraint_->GetBlock(0, 1)));
+            J_21_ = std::unique_ptr<mfem::HypreParMatrix>(
+                static_cast<mfem::HypreParMatrix*>(&J_constraint_->GetBlock(1, 0)));
+            J_22_ = std::unique_ptr<mfem::HypreParMatrix>(
+                static_cast<mfem::HypreParMatrix*>(&J_constraint_->GetBlock(1, 1)));
+
+            // eliminate bcs and compute eliminated blocks
+            J_e_    = bcs_.eliminateAllEssentialDofsFromMatrix(*J_);
+            J_e_21_ = std::unique_ptr<mfem::HypreParMatrix>(J_21_->EliminateCols(bcs_.allEssentialTrueDofs()));
+            J_12_->EliminateRows(bcs_.allEssentialTrueDofs());
+
+            // create block operator for constraints
+            J_constraint_e_ = std::make_unique<mfem::BlockOperator>(J_offsets_);
+            J_constraint_e_->SetBlock(0, 0, J_e_.get());
+            J_constraint_e_->SetBlock(1, 0, J_e_21_.get());
+
+            J_operator_ = J_constraint_.get();
+            return *J_constraint_;
+          });
+    } else {
+      return std::make_unique<mfem_ext::StdFunctionOperator>(
```

Comment:
Just reading quickly, i dont understand this case.  is it penalty, or some case where there are no multipliers on processor?  Maybe create two functions for the 2 cases with names making it clear what the two cases are doing?

Reply:
Added some comments to clarify what each case does